File: go/0101_symmetric_tree.go
package main

func isSymmetric(root *TreeNode) bool {
	if root == nil {
		return true
	}
	nodes := []*TreeNode{root}
	dfsRight(&nodes, root)

	res, count := true, 0
	dfsLeft(&res, &count, nodes, root)
	return res
}

func dfsRight(nodes *[]*TreeNode, root *TreeNode) {
	if root == nil {
		return
	}
	*nodes = append(*nodes, root.Right)
	dfsRight(nodes, root.Right)
	*nodes = append(*nodes, root.Left)
	dfsRight(nodes, root.Left)
}

func dfsLeft(res *bool, count *int, nodes []*TreeNode, root *TreeNode) {
	if root == nil {
		return
	}

	*count = *count + 1
	node, left := nodes[*count], root.Left
	if (node == nil && left == nil) || (node != nil && left != nil && node.Val == left.Val) {
		dfsLeft(res, count, nodes, root.Left)
	} else {
		*res = false
		return
	}

	*count = *count + 1
	node, right := nodes[*count], root.Right
	if (node == nil && right == nil) || (node != nil && right != nil && node.Val == right.Val) {
		dfsLeft(res, count, nodes, root.Right)
	} else {
		*res = false
		return
	}
}

File: python/0101_symmetric_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isSymmetric(self, root: TreeNode) -> bool:
        if not root:
            return True

        return self.traverse(root.left, root.right)

    def traverse(self, left, right):
        if not left and not right:
            return True
        elif (not left and right) or (left and not right) or (left.val != right.val):
            return False

        return self.traverse(left.left, right.right) and self.traverse(left.right, right.left)